src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,586 @@
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 0);
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cached: false, Callback: stored
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithStoredCallback) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillOnce(Invoke(&mock_quota_transport_,
+                       &MockQuotaTransport::AllocateQuotaWithStoredCallback));
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; });
+
+  EXPECT_EQ(done_status, Status::UNKNOWN);
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 1);
+
+  // execute stored callback
+  mock_quota_transport_.on_done_vector_[0](done_status.OK);
+  EXPECT_EQ(done_status, Status::OK);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cached: false, Callback: stored
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithStoredCallbackMultipleReqeusts) {
+  // Initialize client instance with cache disabled
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(0, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.quota_transport = mock_quota_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  std::unique_ptr<ServiceControlClient> client =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(
+          Invoke(&mock_quota_transport_,
+                 &MockQuotaTransport::AllocateQuotaWithStoredCallback));
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  // call Quota 10 times
+  for (int i = 0; i < 10; i++) {
+    client->Quota(quota_request1_, &quota_response,
+                  [&done_status](Status status) { done_status = status; });
+    EXPECT_EQ(done_status, Status::UNKNOWN);
+  }
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 10);
+
+  // execute stored callback
+  mock_quota_transport_.on_done_vector_[0](done_status.OK);
+  EXPECT_EQ(done_status, Status::OK);
```

Comment:
There are 10 in the vector, why only call the first one?  should we call all 10 on_done?   maybe we should have a vector done_status, one for each.